### PR TITLE
#171380771 Travel admins create accommodations

### PR DESCRIPTION
**Pull request by descholar-ceo**
### What does this PR ​ do​ ? 
Allows travel admins, suppliers to create accommodation facilities

### Description of Task to be completed?

Have the following APIs ready and functioning:

`POST /api/accommodations`

`GET /api/accommodations` or `GET /api/accommodations?city=cityName&&page=pageNumber`

`POST /api/accommodations/:id/rooms`

`PATCH /api/accommodations`

`PATCH /api/accommodations/rooms/:id`

`DELETE /api/accommodations/:id`

`DELETE /api/accommodations/rooms/:id`

##### A.Pre-requisites : 
* To be able to test this PR locally, you need to have [PostgreSQL version `9.5.21` or higher](https://www.postgresql.org/) database engine already installed on your local machine or running in a container.
* You need to have [NodeJS version between `10` and 13.4.0](https://nodejs.org/en/) and [npm version `6` or higher](https://www.npmjs.com/)
* You need to have [postman](https://www.postman.com/downloads/) installed on your local machine to be able to test the endpoint
* Your local machine needs to be able to run terminal, powershell, or bash commands

##### B.Steps:
1. If you don't yet have this repository on your local machine, clone it with this command in your terminal `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend`
2. After cloning the repository, run the following command respectively
    > * cd stackup2-barefoot-backend
    > * git fetch origin ft-admin-create-accomodation-171380771
    > * git checkout ft-admin-create-accomodation-171380771
    > * npm install
3. Rename the file `.env.example` to `.env` which is in the root directory of this project (current directory if you correctly followed all of the above steps respectively)
4. Fill the values of PORT, and DEVELOPMENT DATABASE VALUES
5. Then run the following commands in your terminal respectively:
    > * npm run create-db #if the DB you added in `.env` file doesn't exist
    > * npm run migrations
    > * npm start
6. Provide values to the following fields 
```javascript
#SUPER USER 
SUPER_ADMIN_PASSWORD=
SUPER_ADMIN_EMAIL=
```
 in your .env file, and run `npm run admin`, Signup to create a new account, assign role of `travel admin using that super-user` [refer here how to assign role](https://github.com/Stackup-Rwanda/stackup2-barefoot-backend/pull/35), verify your account and log in, and paste the token you get in headers as authorization `Bearer token`

7. After all of the previous steps open your postman and

7.i.    send POST request to `/api/accommodations` to create an accommodation facility by sending the following fields in formData of postman
Screenshot from 2020-04-23 21-44-51

7.ii.    and send another POST request to `/api/accommodations/1/room` with the following fields 

```javascript
{
	"roomNumber":"100B",
	"roomType":"double-room"
}
```
===> That will add room to the accommodation you have just created

7.iii.    send GET request to `/api/accommodation` to see a list of accommodations available or `GET    /api/accommodation?city=kigali` to get list of accommodations in kigali only

7.iv.    send a PATCH request to `/api/accommodations/1` to update the accommodation 1, you can edit anything you want

7.v.    send a PATCH request to `/api/accommodations/rooms/1` to edit room 1

7.vi.    send a DELETE request to `/api/accommodations/1` to delete accommodation 1

7.vii.    send a DELETE request to `/api/accommodations/rooms/1` to delete room 1

### Any background context you want to provide?
If you are running [Node 14.x](https://nodejs.org/en/), this app might behave strangely, unlikely to work for you

### What are the relevant pivotal tracker stories?
[ #171380771](https://www.pivotaltracker.com/n/projects/2435482/stories/171380771)
[ #171380775](https://www.pivotaltracker.com/n/projects/2435482/stories/171380775)

#### 1. When an accommodation is created successfully
Screenshot from 2020-04-13 19-40-28
#### 2. When the accommodation-room is added successfully
Screenshot from 2020-04-13 19-42-48

#### 3. When the accommodations are retrieved successfully
Screenshot from 2020-04-13 19-44-08

### Questions:
N/A

### CODE REVIEW

**Review by Ugizwenayo-Divine — COMMENTED** [review]
Great work Emmanuel but I think you should check that
In custom messages instead of  "There are no any accommodations found in the city you entered!"
you can do " There are no accommodations in the city you entered!"
same as here "There are no any accommodations found on this page!".

File: src/database/models/accommodation.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+export default (sequelize, DataTypes) => {
+  const accommodation = sequelize.define('accommodation', {
+    userId: DataTypes.INTEGER,
+    myImage: DataTypes.STRING,
+    accommodationName: DataTypes.STRING,
+    accommodationAddress: DataTypes.STRING,
+    accommodationDescription: DataTypes.TEXT,
+    cost: DataTypes.STRING,
```
Comment (Muhire-Josue) [inline src/database/models/accommodation.js]: can make this property integer, and create the `currency` property to hold the currency, so in the response, the cost property will be a combination of cost and currency, i.e const cost =  `  `${cost} ${currency}` `

File: tests/other/zaccommodation.test.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import chai, { expect } from 'chai';
```
Comment (Muhire-Josue) [inline tests/other/zaccommodation.test.js]: can you rename this test file? `zaccommodation.test.js` is not really descriptive, please find a descriptive name.

**Review by Muhire-Josue — DISMISSED** [review]
You did a great job @descholar-ceo, really it's fantastic, however, I left a few comments also please resolve those conflicts.

File: src/database/models/accommodation.js (hunk 1 above)
Comment (descholar-ceo) [reply src/database/models/accommodation.js]: This comment of yours is fixed by now

File: tests/other/zaccommodation.test.js (hunk 2 above)
Comment (descholar-ceo) [reply tests/other/zaccommodation.test.js]: I did this for the sake of other previous tests which the accommodations tests depends on

**Comment by descholar-ceo** [comment]
> You did a great job @descholar-ceo, really it's fantastic, however, I left a few comments also please resolve those conflicts.

Yes @Muhire-Josue I have resolved the conflicts already

**Review by since9629 — COMMENTED** [review]
@de-scholar  After going through your PR, I have come up with some suggestions:
- Could you use plural versions of endpoints (e.g `/api/accommodation` => `/api/accommodations`). You can refer to how other endpoints are implemented in our API
- Could you use `city` instead of just `address`, because the former is more descriptive than the later. So `GET /api/accommodation?address=cityName` becomes `GET /api/accommodations?city=cityName`
- Could you change `POST /api/accommodation/accommodation-room` to POST /api/accommodations/rooms , because the former is more descriptive than the later.
- I wonder if we need these fields; `cost`, `status`, `amenities `and `services` since there is no mention of any of them in your story.
- Could you validate user input
- Since there is no way to use enums in es6, I would suggest that you use constants for `roomType `(e.g `DOUBLE_ROOM `instead of plain "double-room")
- Could you stick to the object structure returned while fetching data
```
{
  message: <message>,
  data: <rows>,
  count: <count of data/rows> (only add this property)
}
```

**Review by Ugizwenayo-Divine — APPROVED** [review]
well done

**Review by since9629 — APPROVED** [review]

File: src/database/migrations/20200310215520-create-user.js [hunk 3]
```diff
@@ -74,7 +76,7 @@ export default {
     phoneNumber: {
       type: Sequelize.STRING
     },
-    profileImage: {
+    myImage: {
```
Comment (kipropbrian) [inline src/database/migrations/20200310215520-create-user.js]: Dont edit this migration

**Review by kipropbrian — COMMENTED** [review]
- dont delete the old migrations. like `src/database/migrations/20200415191041-create-accommodation.js`.

**Comment by kipropbrian** [comment]
- accommodationAddress or name does not allow spaces. - I can't put two words with a space in between
- Super admin cannot create accomodations? Why not. 
- One of room name or accomodation should be allowed to be empty. I suggest name.
- Why is accomodation Image called `myImage`? Can I upload an image?
- When editing a room. The response is not formatted well.  Same for editing an accomodation.

```
{
    "message": null,
    "token": null,
    "data": [
        1,
        [
            {
                "id": 1,
                "accommodationId": 1,
                "roomName": "45B",
```
- The response from `GET ROOM` does not have `roomNumber`. 
- Error when deleting an accomadation.
``` DELETE FROM "accommodations" WHERE "id" = 1
(node:97283) UnhandledPromiseRejectionWarning: SequelizeForeignKeyConstraintError: update or delete on table "accommodations" violates foreign key constraint "accommodationRooms_accommodationId_fkey" on table "accommodationRooms"
```

**Comment by kipropbrian** [comment]
There is an error when uploading images 

/Users/brianmaiyo/dev/andela/stackup2-barefoot-backend/src/utils/profile.utils.js:669
    req.body.myImage = uploadResult.url;

TypeError: Cannot read property 'url' of undefined
    at /Users/brianmaiyo/dev/andela/stackup2-barefoot-backend/src/utils/profile.utils.js:24:41
    at /Users/brianmaiyo/dev/andela/stackup2-barefoot-backend/node_modules/cloudinary/lib/utils/index.js:1046:14

Could you check on that?

**Review by kipropbrian — APPROVED** [review]

**Pull request merged**